packages/aws-cdk/lib/api/garbage-collection/stack-refresh.ts

```diff
@@ -0,0 +1,180 @@
+      const template = await cfn.getTemplate({
+        StackName: stack,
+      }).promise();
+
+      templates.push((template.TemplateBody ?? '') + JSON.stringify(summary?.Parameters));
+    }
+  }
+
+  debug('Done parsing through stacks');
+
+  return templates;
+}
+
+/**
+ * Filter out stacks that we KNOW are using a different bootstrap qualifier
+ * This is mostly necessary for the integration tests that can run the same app (with the same assets)
+ * under different qualifiers.
+ * This is necessary because a stack under a different bootstrap could coincidentally reference the same hash
+ * and cause a false negative (cause an asset to be preserved when its isolated)
+ * This is intentionally done in a way where we ONLY filter out stacks that are meant for a different qualifier
+ * because we are okay with false positives.
+ */
+function bootstrapFilter(parameters?: CloudFormation.ParameterDeclarations, qualifier?: string) {
+  const bootstrapVersion = parameters?.find((p) => p.ParameterKey === 'BootstrapVersion');
+  const splitBootstrapVersion = bootstrapVersion?.DefaultValue?.split('/');
+  // We find the qualifier in a specific part of the bootstrap version parameter
+  return (qualifier &&
+          splitBootstrapVersion &&
+          splitBootstrapVersion.length == 4 &&
+          splitBootstrapVersion[2] != qualifier);
+}
+
+export async function refreshStacks(cfn: CloudFormation, activeAssets: ActiveAssetCache, maxWaitTime: number, qualifier?: string) {
+  try {
+    const stacks = await fetchAllStackTemplates(cfn, maxWaitTime, qualifier);
+    for (const stack of stacks) {
+      activeAssets.rememberStack(stack);
+    }
+  } catch (err) {
+    throw new Error(`Error refreshing stacks: ${err}`);
+  }
+}
+
+/**
+ * Background Stack Refresh properties
+ */
+export interface BackgroundStackRefreshProps {
+  /**
+   * The CFN SDK handler
+   */
+  readonly cfn: CloudFormation;
+
+  /**
+   * Active Asset storage
+   */
+  readonly activeAssets: ActiveAssetCache;
+
+  /**
+   * Stack bootstrap qualifier
+   */
+  readonly qualifier?: string;
+
+  /**
+   * Maximum wait time when waiting for stacks to leave REVIEW_IN_PROGRESS stage.
+   *
+   * @default 60000
+   */
+  readonly maxWaitTime?: number;
+}
+
+/**
+ * Class that controls scheduling of the background stack refresh
+ */
+export class BackgroundStackRefresh {
+  private timeout?: NodeJS.Timeout;
+
+  constructor(private readonly props: BackgroundStackRefreshProps) {}
+
+  public async start() {
+    const startTime = Date.now();
+
+    await refreshStacks(this.props.cfn, this.props.activeAssets, this.props.maxWaitTime ?? 60000, this.props.qualifier);
+
+    // If the last invocation of refreshStacks takes <5 minutes, the next invocation starts 5 minutes after the last one started.
+    // If the last invocation of refreshStacks takes >5 minutes, the next invocation starts immediately.
+    this.timeout = setTimeout(this.start, Math.max(startTime + 300_000 - Date.now(), 0));
```

Comment:
This will not do what you want, because of `this` in JavaScript. It's not you, it would have been correct in all other sane languages, just not JS.

You need the following:

```suggestion
    this.timeout = setTimeout(this.start.bind(this), Math.max(startTime + 300_000 - Date.now(), 0));
```

Or this:

```suggestion
    this.timeout = setTimeout(() => this.start(), Math.max(startTime + 300_000 - Date.now(), 0));
```

Whichever one you think is more readable.